src/box/iproto.cc

```diff
@@ -504,6 +507,26 @@ struct iproto_connection
 	 * connection.
 	 */
 	struct cmsg destroy_msg;
+	/**
+	 * Pre-allocated msg. Is sent after shutdown has started. If session
+	 * support graceful shutdown, then pushes graceful_shutdown_msg to its
+	 * route, else pushes not_graceful_shutdown_msg to its route.
+	 */
+	 struct cmsg shutdown_msg;
+	/**
+	 * Pre-allocated msg. Is sent after shutdown has started only if
+	 * session support graceful shutdown. Sends IPROTO_SHUTDOWN to remote
+	 * client.
+	 */
+	struct cmsg graceful_shutdown_msg;
+	/**
+	 * Pre-allocated msg. Is sent after shutdown has started only if
+	 * session doesn't support graceful shutdown(or isn't exist). Reads
+	 * data from socket until iproto boundary or until msg_max_limit is
+	 * reached and shutdown(con, SHUT_RD). Then, notify session (if session
+	 * exists) about shutdown of socket.
+	 */
+	struct cmsg not_graceful_shutdown_msg;
```

Comment:
The amount of preallocated messages is getting out of hand IMO. AFAIU, only one of them is used, so you could make it one, and in the target function check which type of the shutdown that is. Or could even try to merge these messages together including destroy_msg and create a single "service route" for all service messages.